crypto/x509/x509_vfy.c

```diff
@@ -833,41 +830,26 @@ static int check_trust(X509_STORE_CTX *ctx, int num_untrusted)
     }
 
     /*
-     * If we are looking at a trusted certificate, and accept partial chains,
-     * the chain is PKIX trusted.
+     * If there are no new trusted certificates,
+     * check the current cert for a direct trust store match
      */
-    if (num_untrusted < num) {
-        if (ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN)
-            goto trusted;
-        return X509_TRUST_UNTRUSTED;
-    }
-
-    if (num_untrusted == num && ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN) {
-        /*
-         * Last-resort call with no new trusted certificates, check the leaf
-         * for a direct trust store match.
-         */
-        i = 0;
-        x = sk_X509_value(ctx->chain, i);
-        mx = lookup_cert_match(ctx, x);
-        if (!mx)
-            return X509_TRUST_UNTRUSTED;
+    x = sk_X509_value(ctx->chain, num_untrusted - 1);
+    if (num_untrusted == num
+            && (mx = lookup_cert_match(ctx, x)) != NULL) {
```

Comment:
This needs to be interpreted in the context of the above mentioned core change of this PR, namely adding to `build_chain()`:
```
/* Stop with success if current cert equals a trust anchor */
            if (lookup_cert_match(ctx, x) != NULL
                    && (trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED)
                break; 
 ```
 If there a direct match of the current cert (i.e., the current end of the chain being built) with a trusted cert (no matter if it's a not self-signed EE cert or any other sort of cert) then we can succeed immediately.
Maybe the generality of having potentially some intermediate certs before is not needed anymore?

Reply:
I've had a look at this again.
The 'last resort' case (for an EE cert that is in the trust store) that you mention above:
> directly match an EE cert that is not self-signed and comes along with an untrusted issuer chain.

is covered by the new, generalized code because due to the new call `(trust = check_trust(ctx, num)) == X509_TRUST_TRUSTED` in `build_chain()`, `check_trust()` gets called for each trust store element added to the chain, and  when called for the first time `num_untrusted` will be `1`, so the EE (leaf) cert will be handled as desired.
This is witnessed by the test cases `"accept last-resort direct leaf match"` which still pass.

Note that in the PKIX case the check commented as `Last chance to make a trusted chain,
either bare DANE-TA public-key signers or direct leaf PKIX trust.`
after the loop is now not needed anymore, so I've removed it.

Moreover, the code section in build_chain()`:
```
                } else if (num == ctx->num_untrusted) {
                    /*
                     * We have a self-signed certificate that has the same
                     * subject name (and perhaps keyid and/or serial number) as
                     * a trust anchor.  We must have an exact match to avoid
                     * possible impersonation via key substitution etc.
                     */
                    if (X509_cmp(curr, issuer) != 0) {
                        /* Self-signed untrusted mimic. */
                        X509_free(issuer);
                        ok = 0;
                    } else { /* curr "==" issuer */
                        X509_free(curr);
                        ctx->num_untrusted = --num;
                        (void)sk_X509_set(ctx->chain, num, issuer);
                        curr = issuer;
                        /* no need to update self_signed */
                    }
                }

```
was used only in two degenerate cases:
 * for direct trust of EE cert
 * if X509_V_FLAG_TRUSTED_FIRST unset and trust anchor is also in untrusted

I've disabled it as well because it is covered by the generalized `check_trust()` now called before,
which checks direct trust for current cert (at index num_trusted - 1).

All this did not lead to a change in documented behavior.